### ISSUE TITLE
Redesign the commit stage

### ISSUE THREAD

**Issue by crusaderky** [deshaw/versioned-hdf5#398]
Downstream of #386, the commit stage - when you exit the `stage_version(...)` context manager - is by far what takes the longest time. This is unsurprising, as #386 didn't touch it.

We should extend the redesign of #386 to everything that happens afterwards, which means hashing and writing to disk.
The StagedChangesArray introduced in #386 offers solid infrastructure to build on.

# Phases
This rewrite can happen in separate phases, each with an incremental benefit to production:
1. push into the StagedChangesArray the hashing phase and the writing to hdf5 when writing from a InMemoryDataset or InMemorySparseDataset, leaving the creation of the virtual dataset to the legacy code;
2. rewrite the creation of the virtual dataset downstream of InMemoryDataset or InMemorySparseDataset in a way that natively interfaces with the metadata of the StagedChangesArray, can be much more aggressively cythonized, and doesn't use the data_dict nor ndindex at any point;
3. rewrite create_dataset and all other machinery on top of the StagedChangesArray so that the legacy code can be retired.

# Performance expectations
Committing a 3 GiB dataset with 40,000x 80 kiB chunks (all chunks modified and different from the previous values) takes

- 12.7s today (as measured on NVMe Linux desktop)
- ~9.1s expected after phase 1
- ~5.8s expected after phase 2

A granular breakdown of these numbers appers later in this document.

# Low level design

**NOTE:** this section expects the reader to be already familiar with the low level design of #386. If you aren't, please stop here and read https://github.com/deshaw/versioned-hdf5/blob/17c687d3e1f1bfb64781a7b9039dbb621f5a46b4/docs/staged_changes.rst first.

Create a new `StagedChangesArray.hash_tables` list, matching the `slabs` list.
Every slab has a matching hash_table with dtype=hsize_t and shape=(n, 4), where n is the number of chunks in the slab (so every row can contain exactly one sha256 hash).

`hash_tables[i][j, :]` is the hash for the chunk on the slab at `slabs[i][j * chunk_size[0]:(j + 1) * chunk_size[0]]`.

New requirement: base slabs must be aligned to `chunk_sizes[0]`. This is de facto already true for the `raw_data` slab.

New hash tables are created with `np.zeros` together with the matching slabs. A hash set to 0 means that it's uninitialised.

The hash_tables entry for the raw_data base slab is an actual np.ndarray loaded from disk, not a h5py.Dataset.

### Committing changes
To "commit" is the act of taking all chunks that lie on a staged slab and moving them to a new base slab. The new base slab is created by a callback that must replicate the API of numpy.empty. InMemoryDataset, instead of creating a new empty numpy array, will enlarge raw_data and then return an ad-hoc view to it (using a thin wrapper class).

Committing is broken down in two stages: HashPlan and CommitPlan.

### HashPlan
This new very simple plan finds all staged chunks and passes the sorted list of slab indices and slab offsets to a cythonized sha256 hasher, which will feature a vectorized API very similar to `read_many_slices`. 

`StagedChangesArray.commit()` calls the hasher:
```python
    for slab_idx, instructions in plan.hashes:
       slab = self._get_slab(slab_idx)
       hash_table = self._get_hash_table(slab_idx)
       hasher(slab, hash_table, instructions)  # pure C
```

The hasher populates `StagedChangesArray.hash_tables` for all staged chunks (but, crucially, not for deleted chunks and without covering uninitialised memory that may be on the edge chunks).
It also calculates the hash of the full chunk, as if it were C-contiguous.

### CommitPlan
This new plan takes in input the full `StagedChangesArray.hash_tables` list.

First, it scans hash_tables and uses it to compile a hashmap of  `{hash key: (slab idx, slab offset)}`.

Slabs with lower numeration overwrite the ones with higher numeration, meaning that
- if a real chunk is actually full of `fill_value`, it will be replaced by the full chunk
- if a chunk on a staged slab is identical to one on a base slab, it will be replaced by the one on the base slab.

*All* hashes are parsed - not just the ones referenced by a chunk - so that a chunk that was committed 2+ versions ago and then was deleted, but still exists in raw_data, it can be reused in the coming commit.

Once the map is compiled, we can once again scan for all staged chunks. This query is identical to the one we used in the HashPlan moments ago and the previous output should be reused for efficiency.

For each chunk, we now proceed to update slab_indices and slab_offsets with whatever we find in the hashmap (slab with lowest index wins). We have successfully deduplicated our staged chunks among each other and against the base chunk. Whenever a staged chunk already exists on the base slab, it now becomes just a reference to it.

Now we *insert* the new base slab just after the other base slabs. This is different from what we do normally, which is append at the end of the slabs list. CommitPlan itself will also create a new hash_tables entry, unlike the other MutatingPlans that just instruct StagedChangesArray to do so.

We scan slab_indices and slab_offsets again. Same query as before, but different result because of the deduplication pass we just completed.

We then compile a TransferPlan. This is the inverse of LoadPlan, where all the contents of the base slabs are copied to a new staged slab and then dropped. Here, we copy all the contents of the staged slabs (those that survived deduplication) to a new base slab and finally we drop all the staged slabs.

### From InMemoryDataset/InMemorySparseDataset's POV

1. Commit starts: there is 1 base slab, which is raw_data, and 0 or more staged slabs.
2. Call `self.staged_changes.commit(empty=self.resize_raw_data_and_return_view)` where the callback resizes `self.id.raw_data` and returns a view to the new surface.
3. There are now exactly 3 elements in `self.staged_changes.slabs`:
    - `slabs[0]`: the full slab (ignore it)
    - `slabs[1]`: `raw_data` (may have been replaced with None)
    - `slabs[2]`: view to extended `raw_data` we just created with the callback.

All data - but not the metadata yet - has been saved to disk.

4. We can now resize the hash table in the hdf5 dataset and append to it `staged_changes.hash_tables[2]`.
5. finally, we hack the metadata:
```python
   self.staged_changes.slab_offsets[
       self.staged_changes.slab_indices == 2
   ] += previous_raw_data_len
```

## Recreating the virtual dataset
At the end of the last stage we can transitionally resume with what we were doing in the legacy design: call `self.staged_changes.changes()`, compile a `data_dict`, and pass it to `create_virtual_dataset`.

This also has the advantage that we don't need to think yet about what happens in the `create_dataset` function.

In a later phase, however, we can rewrite `create_virtual_dataset` so that it accepts `slab_indices` and `slab_offsets` directly, in a way that mirrors `build_slab_indices_and_offsets` and doesn't require a data_dict or ndindex.
When this happens, the `data_dict` and the `CommitPlan` can be retired.

`create_dataset` will need to be reimplemented on top of `StagedChangesArray.from_array(as_base=False)`, which will create staged slabs from views of the original data.

#### Phase 1:
1. Add hash table to InMemoryDatasetID, next to raw_data
2. staged_changes.py: StagedChangesArray.commit(), HashPlan, CommitPlan 
3. Write cythonized hasher
4. integration in InMemoryDataset (resize raw_data; append to hashes)

#### Phase 2:
5. Rewrite create_virtual_dataset

#### Phase 3:
6. Rewrite new_dataset on top of StagedChangesArray (needed to delete legacy code)

## Benchmarks
Breakdown of `vf.stage_version(...).__exit__()` to commit a dataset with
- `shape=(20_000, 20_000)` (3 GiB)
- `chunks=(100, 100)` (40_000 chunks worth 80 kiB each)

and all chunks modified (write to local NVMe):
TODAY (as of #386)
All timings are measured with pyinstrument:
```
 1.3s InMemoryDataset.data_dict
 1.6s openssl_sha256
 0.9s Hashtable control code
 5.7s Dataset.__setitem__
 2.7s create_virtual_dataset
 0.5s misc control code
12.7s TOTAL
```
After phase 1 (move hashing and writing to StagedChangesArray).
```
 0.4s plans formulation (guesstimate)
 1.6s hash_slabs (measured on POC code)
 3.1s read_many_slices calls H5Dwrite for each chunk (measured on POC code)
 1.3s InMemoryDataset.data_dict (same as prod)
 2.7s create_virtual_dataset (same as prod)
 9.1s TOTAL
```
_Note:_ Early POC benchmarks show no benefit of calling openssl C hashers directly vs. hashlib at 10k points per chunk (80kb), and start seeing a substantial benefit at 1000 points per chunk (8 kb).

After phase 2 (rewrite create_virtual_dataset):
```
 0.4s plans formulation (guesstimate)
 1.6s hash_slabs (measured on POC code)
 3.1s read_many_slices calls H5Dwrite for each chunk (measured on POC code)
 0.7s create_virtual_dataset (guesstimate)
 5.8s TOTAL
```

# Out of scope bonus: cache on read
Downstream of this it becomes straightforward, if we wish to, to implement a switch that enables cache on read for __getitem__, like it used to happen before #386.

GetItemPlan becomes a MutatingPlan with the option, if the cache-on-read flag is on, to load a new staged slab from hdf5 and then read from it, just like SetItemPlan loads a new staged slab from hdf5 for the partially modified data and then updates it.

When GetItemPlan creates a new slab, it also appends a new array to hash_tables and populates it with a copy of the hashes it just copied. This means that HashPlan will later have an easy time skipping the hashing of the array. 

LoadPlan likewise copies over the hashes from the base slabs to the staged slabs.

When SetItemPlan and ResizePlan update an existing slab, they instead zero out the updated chunks to signal that the hash needs to be (re)computed.
astype() completely wipes the hash tables.
refill() could partially wipe the hash tables only where it actually changes points, but for the sake of simplicity it's likely simpler to just wipe everything.

**Comment by ArvidJB** [comment]
@crusaderky: let's discuss this in our meeting on Wednesday.

**Comment by ArvidJB** [comment]
Here's a profile:

<img width="719" height="706" alt="Image" src="https://github.com/user-attachments/assets/0828abb8-d9b6-48b1-b4a6-4a5fdd1e0ae7" />

This is for writing to an HDF5 file with many large arrays:
```
column 0: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 1: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 2: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 3: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 4: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 5: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 6: shape=(40225637,) dtype=float64 raw_data.shape=(42810000,), raw_data.chunks=(10000,), hash_table.shape=(4281,)
column 7: shape=(40225637,) dtype=float64 raw_data.shape=(42810000,), raw_data.chunks=(10000,), hash_table.shape=(4281,)
column 8: shape=(40225637,) dtype=float64 raw_data.shape=(10480000,), raw_data.chunks=(10000,), hash_table.shape=(1048,)
column 9: shape=(40225637,) dtype=float64 raw_data.shape=(10480000,), raw_data.chunks=(10000,), hash_table.shape=(1048,)
column 10: shape=(40225637,) dtype=float64 raw_data.shape=(27840000,), raw_data.chunks=(10000,), hash_table.shape=(2784,)
column 11: shape=(40225637,) dtype=float64 raw_data.shape=(27850000,), raw_data.chunks=(10000,), hash_table.shape=(2785,)
column 12: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 13: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 14: shape=(40225637,) dtype=float64 raw_data.shape=(24020000,), raw_data.chunks=(10000,), hash_table.shape=(2402,)
column 15: shape=(40225637,) dtype=float64 raw_data.shape=(24030000,), raw_data.chunks=(10000,), hash_table.shape=(2403,)
column 16: shape=(40225637,) dtype=float64 raw_data.shape=(38330000,), raw_data.chunks=(10000,), hash_table.shape=(3833,)
column 17: shape=(40225637,) dtype=float64 raw_data.shape=(38330000,), raw_data.chunks=(10000,), hash_table.shape=(3833,)
column 18: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 19: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 20: shape=(40225637,) dtype=float64 raw_data.shape=(28700000,), raw_data.chunks=(10000,), hash_table.shape=(2870,)
column 21: shape=(40225637,) dtype=float64 raw_data.shape=(28710000,), raw_data.chunks=(10000,), hash_table.shape=(2871,)
column 22: shape=(40225637,) dtype=float64 raw_data.shape=(40350000,), raw_data.chunks=(10000,), hash_table.shape=(4035,)
column 23: shape=(40225637,) dtype=float64 raw_data.shape=(40350000,), raw_data.chunks=(10000,), hash_table.shape=(4035,)
column 24: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 25: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 26: shape=(40225637,) dtype=float64 raw_data.shape=(14140000,), raw_data.chunks=(10000,), hash_table.shape=(1414,)
column 27: shape=(40225637,) dtype=float64 raw_data.shape=(14140000,), raw_data.chunks=(10000,), hash_table.shape=(1414,)
column 28: shape=(40225637,) dtype=float64 raw_data.shape=(32450000,), raw_data.chunks=(10000,), hash_table.shape=(3245,)
column 29: shape=(40225637,) dtype=float64 raw_data.shape=(32460000,), raw_data.chunks=(10000,), hash_table.shape=(3246,)
column 30: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 31: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 32: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 33: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 34: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 35: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 36: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 37: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 38: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 39: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 40: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 41: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 42: shape=(40225637,) dtype=float64 raw_data.shape=(130000,), raw_data.chunks=(10000,), hash_table.shape=(13,)
column 43: shape=(40225637,) dtype=float64 raw_data.shape=(130000,), raw_data.chunks=(10000,), hash_table.shape=(13,)
column 44: shape=(40225637,) dtype=float64 raw_data.shape=(1300000,), raw_data.chunks=(10000,), hash_table.shape=(130,)
column 45: shape=(40225637,) dtype=float64 raw_data.shape=(1300000,), raw_data.chunks=(10000,), hash_table.shape=(130,)
column 46: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 47: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 48: shape=(40225637,) dtype=float64 raw_data.shape=(980000,), raw_data.chunks=(10000,), hash_table.shape=(98,)
column 49: shape=(40225637,) dtype=float64 raw_data.shape=(980000,), raw_data.chunks=(10000,), hash_table.shape=(98,)
column 50: shape=(40225637,) dtype=float64 raw_data.shape=(1450000,), raw_data.chunks=(10000,), hash_table.shape=(145,)
column 51: shape=(40225637,) dtype=float64 raw_data.shape=(1450000,), raw_data.chunks=(10000,), hash_table.shape=(145,)
column 52: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 53: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 54: shape=(40225637,) dtype=float64 raw_data.shape=(3050000,), raw_data.chunks=(10000,), hash_table.shape=(305,)
column 55: shape=(40225637,) dtype=float64 raw_data.shape=(3050000,), raw_data.chunks=(10000,), hash_table.shape=(305,)
column 56: shape=(40225637,) dtype=float64 raw_data.shape=(4220000,), raw_data.chunks=(10000,), hash_table.shape=(422,)
column 57: shape=(40225637,) dtype=float64 raw_data.shape=(4220000,), raw_data.chunks=(10000,), hash_table.shape=(422,)
column 58: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 59: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 60: shape=(40225637,) dtype=float64 raw_data.shape=(2600000,), raw_data.chunks=(10000,), hash_table.shape=(260,)
column 61: shape=(40225637,) dtype=float64 raw_data.shape=(2600000,), raw_data.chunks=(10000,), hash_table.shape=(260,)
column 62: shape=(40225637,) dtype=float64 raw_data.shape=(4620000,), raw_data.chunks=(10000,), hash_table.shape=(462,)
column 63: shape=(40225637,) dtype=float64 raw_data.shape=(4620000,), raw_data.chunks=(10000,), hash_table.shape=(462,)
column 64: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 65: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 66: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 67: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 68: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 69: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 70: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 71: shape=(40225637,) dtype=float64 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 72: shape=(40225637,) dtype=|S1 raw_data.shape=(41070000,), raw_data.chunks=(10000,), hash_table.shape=(4107,)
column 73: shape=(40225637,) dtype=|S10 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 74: shape=(40225637,) dtype=|S6 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
column 75: shape=(40225637,) dtype=|S16 raw_data.shape=(42880000,), raw_data.chunks=(10000,), hash_table.shape=(4288,)
```

**Comment by crusaderky** [comment]
I  estimate
- 40 wds for phase 1
- 15 wds for phase 2
- 15 wds for phase 3

**Comment by ArvidJB** [comment]
Thanks @crusaderky!

Another question: in your performance expectations you are assuming that *all chunks were modified*. How would the timings change if only a small percentage of the chunks are modified? Wouldn't the performance be substantially better since we don't have to re-compute hashes for chunks which haven't changed? Or is that not a large part of the running time?

**Comment by crusaderky** [comment]
Commit time is equal to

1. (time it takes for numpy to test a single point of `slab_indices`) * (total n. of chunks in the dataset)
2. plus (time it takes to hash a chunk) * (n. of chunks touched by `__setitem__`)
3. plus (time it takes to write a chunk to disk, without one-off overheads) * (n. of unique chunks not yet in `raw_data`)
4. plus (time it takes to add a chunk to the virtual dataset) * (total n. of chunks in the dataset)
5. plus fixed overheads (resize `raw_data`, perform a single write to it not counting the actual bytes, create the vds not counting the actual chunks)

This holds true for both the current and the proposed implementation.
If you change only a single chunk out of a huge dataset, the only costs will be 
- (n.1 above) `_chunks_in_selection` scanning the whole `slab_indices`
- (n.4 above) `create_virtual_dataset`
- (n.5 above) fixed overheads

with n.4 by far dominating the total runtime.

**Comment by ArvidJB** [comment]
So if we did implement the proposal, can you "guesstimate" the expected runtime for your example (shape=(20_000, 20_000), chunks=(100, 100)) if we're 
1. updating 1% of the existing data? In particular, let's assume we're only modifying 1% of the existing *chunks*.
2. appending 1% of new data? so new shape would (101, 100), all existing data except for the "last" chunks `(19900:, :)` would not be touched.

**Comment by crusaderky** [comment]
> 1. updating 1% of the existing data? In particular, let's assume we're only modifying 1% of the existing _chunks_.
> 2. appending 1% of new data? so new shape would (101, 100), all existing data except for the "last" chunks `(19900:, :)` would not be touched.

These two use cases are pretty much the same from a performance standpoint. The only thing that enlarging the dataset adds is a deep-copy of the two numpy arrays `StagedChangesArray.slab_indices` and `slab_offsets`, which are worth kilobytes.


I've rerun pyinstrument with versioned-hdf5 2.4.0, h5py 3.16, and hdf5 2.1:

### All chunks change
```
.  │     └─ 12.958 commit_version  versioned_hdf5/versions.py:79
.  │        ├─ 11.013 write_dataset  versioned_hdf5/backend.py:250
.  │        │  ├─ 7.050 Dataset.__setitem__  h5py/_hl/dataset.py:962
.  │        │  ├─ 1.852 Hashtable.hash  versioned_hdf5/hashtable.py:119
.  │        │  ├─ 0.627 [self]  versioned_hdf5/backend.py
.  │        │  ├─ 0.604 NDIndexConstructor.__call__  ndindex/ndindex.py:112
.  │        │  ├─ 0.383 Hashtable.__setitem__  versioned_hdf5/hashtable.py:227
.  │        │  └─ 0.315 ChunkSize.indices  ndindex/chunking.py:100
.  │        └─ 1.868 create_virtual_dataset  versioned_hdf5/backend.py:557
.  │           ├─ 0.503 [self]  versioned_hdf5/backend.py
.  │           ├─ 0.475 select  h5py/_hl/selections.py:19
.  │           ├─ 0.233 Group.create_virtual_dataset  h5py/_hl/group.py:245
.  │           └─ 0.173 Tuple.isempty  ndindex/tuple.py:672
```
## Single chunk changes
```
.  │     └─ 3.358 commit_version  versioned_hdf5/versions.py:79
.  │        ├─ 2.139 create_virtual_dataset  versioned_hdf5/backend.py:557
.  │        │  ├─ 0.561 [self]  versioned_hdf5/backend.py
.  │        │  ├─ 0.559 select  h5py/_hl/selections.py:19
.  │        │  ├─ 0.476 Group.create_virtual_dataset  h5py/_hl/group.py:245
.  │        │  ├─ 0.147 Tuple.isempty  ndindex/tuple.py:672
.  │        │  ├─ 0.087 SimpleSelection.__init__  h5py/_hl/selections.py:228
.  │        │  ├─ 0.073 Slice.__len__  ndindex/slice.py:73
.  │        │  ├─ 0.065 WeakValueDictionary.__setitem__  weakref.py:138
.  │        │  ├─ 0.060 <genexpr>  versioned_hdf5/backend.py:589
.  │        │  └─ 0.050 remove  weakref.py:105
.  │        ├─ 1.044 InMemoryDataset.data_dict  versioned_hdf5/wrappers.py:657
.  │        │  └─ 1.044 _staged_changes_to_data_dict  versioned_hdf5/wrappers.py:1114
.  │        │     ├─ 0.756 NDIndexConstructor.__call__  ndindex/ndindex.py:112
.  │        │     ├─ 0.221 [self]  versioned_hdf5/wrappers.py
.  │        │     └─ 0.066 Tuple.__hash__  ndindex/tuple.py:51
.  │        ├─ 0.110 write_dataset_chunks  versioned_hdf5/backend.py:477
.  │        └─ 0.064 [self]  versioned_hdf5/versions.py
```

From this I can infer that

- after **phase 1**, which revolves around rewriting the hashing phase in a heavily cythonized way, you would get very little to no benefits in the 1% use case: `Hashtable` doesn't even appear in the profile (as expected).
- after **phase 2** you would get major benefits in both the 100% and the 1% use cases:
  - all ndindex overheads would dissappear
  - `InMemoryDataset.data_dict` would not be needed anymore, and be replaced by a much faster cython algorithm with no python involved (above I guessestimated 0.4s).
  - I have good confidence that create_virtual_dataset would go much faster as it would involve just C calls to the hdf5 C API. I guesstimate 0.4s for the new code that would replace  `versioned_hdf5.backend.create_virtual_dataset` and iteratetively call `layout.dcpl.set_virtual` for each chunk. The single call to `h5py.Group.create_virtual_dataset` (0.5s) that happens at the end of versioned_hdf5.backend.create_virtual_dataset` would remain there. 

So as of phase 2 I would guesstimate that the 1% use case would go down from the current 3.4s to ~1.3s.

**Comment by ArvidJB** [comment]
Thanks @crusaderky, that looks promising! Let's implement phase 1 and phase 2, 55 working days approved.

Let's discuss phase 3 when we get close to wrapping up phase 2.
